cmd/check.go

```diff
@@ -121,85 +120,146 @@ func runCheckCmd(cmd *cobra.Command, args []string) {
+		if err != nil {
+			error_helpers.ShowError(ctx, err)
+			continue
 		}
+	}
 
-		durations = append(durations, executionTree.EndTime.Sub(executionTree.StartTime))
+	// set the defined exit code after successful execution
+	exitCode = getExitCode(totalAlarms, totalErrors)
+}
+
+// exportExecutionTree relies on the fact that the given tree is already executed
+func exportExecutionTree(ctx context.Context, namedTree *namedExecutionTree, initData *control.InitData, exportArgs []string) error {
+	statushooks.Show(ctx)
+	defer statushooks.Done(ctx)
+
+	if error_helpers.IsContextCanceled(ctx) {
+		return ctx.Err()
 	}
 
-	if shouldPrintTiming() {
-		printTiming(args, durations)
+	exportMsg, err := initData.ExportManager.DoExport(ctx, namedTree.name, namedTree.tree, exportArgs)
+	if err != nil {
+		return err
 	}
 
 	// print the location where the file is exported if progress=true
 	if len(exportMsg) > 0 && viper.GetBool(constants.ArgProgress) {
 		fmt.Printf("\n")
-		fmt.Println(strings.Join(exportMsg, "\n"))
+		fmt.Println(exportMsg)
 		fmt.Printf("\n")
 	}
 
-	// set the defined exit code after successful execution
-	exitCode = getExitCode(totalAlarms, totalErrors)
+	return nil
+}
+
+// executeTree executes and displays the (table) results of an execution
+func executeTree(ctx context.Context, tree *controlexecute.ExecutionTree, initData *control.InitData) error {
+	checkCtx := createCheckContext(ctx)
+	err := tree.Execute(checkCtx)
+	if err != nil {
+		return err
+	}
+
+	err = displayControlResults(checkCtx, tree, initData.OutputFormatter)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func publishSnapshot(ctx context.Context, executionTree *controlexecute.ExecutionTree, shouldShare bool, shouldUpload bool) error {
+	if error_helpers.IsContextCanceled(ctx) {
+		return ctx.Err()
+	}
+	// if the share args are set, create a snapshot and share it
+	if shouldShare || shouldUpload {
+		statushooks.SetStatus(ctx, "Publishing snapshot")
+		return controldisplay.PublishSnapshot(ctx, executionTree, shouldShare)
+	}
+	return nil
+}
+
+// getExecutionTrees returns a list of execution trees with the names of their export targets
+func getExecutionTrees(ctx context.Context, initData *control.InitData, args ...string) ([]*namedExecutionTree, error) {
+	trees := []*namedExecutionTree{}
+	if initData.ExportManager.HasNamedExport(viper.GetStringSlice(constants.ArgExport)) {
+		// create a single merged execution tree from all arguments
+		executionTree, err := controlexecute.NewExecutionTree(ctx, initData.Workspace, initData.Client, initData.ControlFilterWhereClause, args...)
+		if err != nil {
+			return nil, sperr.WrapWithMessage(err, "could not create merged execution tree")
+		}
+		name := fmt.Sprintf("check.%s", initData.Workspace.Mod.ShortName)
+		trees = append(trees, newNamedExecutionTree(name, executionTree))
+	} else {
+		for _, arg := range args {
+			if error_helpers.IsContextCanceled(ctx) {
```

Comment:
do we need this here?

Reply:
how long does tree creation take?

Reply:
@kaidaguerre 

when a `--where` or `--tag` is passed in, we use the service to filter controls and benchmarks.